### DISCUSSION TITLE
directData map created ok but EquationGen not reading file at 'generateDirectDataInit'

### DISCUSSION THREAD

**Discussion by Karaka** [climateinteractive/SDEverywhere#170] [Q&A]
@chrispcampbell @travisfranck  Gents...most grateful for any thoughts?

My .mdl file was created using Vensim PLE (so directData feature not available) then I edited the file to add directData for one variable (see line 2):
<img width="1165" alt="Screenshot 2021-12-04 at 08 52 47" src="https://user-images.githubusercontent.com/7798808/144703855-12351dbc-ce45-424d-b012-b9e03e581586.png">

'Sde-generate' seems to create my directData map ok:
<img width="2357" alt="Screenshot 2021-12-04 at 09 19 53" src="https://user-images.githubusercontent.com/7798808/144704427-6819934b-9835-4323-ba2b-ea72d37cf48c.png">

From data.xlsx:
<img width="403" alt="Screenshot 2021-12-04 at 09 24 02" src="https://user-images.githubusercontent.com/7798808/144704512-a73c2106-51e6-4a3c-a9e6-6521db2129ab.png">

And the directData map is found in 'EquationGen' at 'generateDirectDataInit':
<img width="2221" alt="Screenshot 2021-12-04 at 09 33 30" src="https://user-images.githubusercontent.com/7798808/144704774-66de6067-3e54-4dd0-9f6b-21422e87ae39.png">

But line 367 leaves 'workbook' undefined so my run terminates with the error:
<img width="1450" alt="Screenshot 2021-12-04 at 09 39 13" src="https://user-images.githubusercontent.com/7798808/144704879-89c19fdb-fd8d-418b-95ff-a8f553a5aae2.png">

**Comment by travisfranck** [comment]
@Karaka Your screenshots are pretty tiny for me, even when I click on them.  

Can you post both your PLE version and your modified MDL version here, if they are shareable?

**Reply by Karaka** [reply]
Thanks for taking a look @travisfranck.
I see what you mean about the images...sorry about that.  Will try to correct.  Meantime here is the PLE version (may be slightly different to original that I worked on as that was going back some) and my modified version.  Yes very shareable.   And I can email the screenshots directly if helpful in the mean time.

Note that I changed the file type to .txt so that the markdown editor would accept the files.

Thank you.

[PLE original - Impact on staff - positions filled by gender with growth time series reference mode.txt](https://github.com/climateinteractive/SDEverywhere/files/7633501/PLE.original.-.Impact.on.staff.-.positions.filled.by.gender.with.growth.time.series.reference.mode.txt)


[Modified PLE GPG mdl.txt](https://github.com/climateinteractive/SDEverywhere/files/7633550/Modified.PLE.GPG.mdl.txt)

**Reply by Karaka** [reply]
hi @travisfranck ... have edited the screenshots in the original post.  hoping these work better!? bon weekend.

**Reply by chrispcampbell** [reply]
@Karaka In your `GPG_spec.json` file, did you include a `directData` section that includes a reference to your xlsx file?  Here's an example (taken from the `directdata` model test in SDE):

```json
{
  "name": "directdata",
  "directData": {
    "?data": "data.xlsx"
  }
}
```

It looks like in your mdl file you have `GET DIRECT DATA('?data.xlsx')`, so I think you need to make sure your "spec.json" file needs to match; I think you might be able to leave the `directData` section empty if you reference it by name in the mdl file already, or maybe just try something like this:

```json
{
  "name": "GPG",
  "directData": {
    "?data.xlsx": "data.xlsx"
  }
}
```

**Reply by Karaka** [reply]
@chrispcampbell   hi Chris.  Thanks for comment.  Here are the GPG_spec.json commands I've been using...I notice that I'm using "?data" as the key whereas you've got "?data.xlsx"....I'll give that a try this eve and report back.  Cheers

{
  "name": "GPG",
  
  "directData": { "?data": "data.xlsx" },

  "inputVars": [
    "_monthly_growth_in_positions",
    "_annual_turnover_white_males",
    "_probability_of_hiring_or_firing_white_male"
  ],
  "outputVars": [
    "_time",
    "_employee_positions",
    "_representation_of_men",
    "_open_positions",
    "_representation_of_women"
  ]
}

**Reply by Karaka** [reply]
Just did a quick run with your format @chrispcampbell and its moved on to generate c code!  So that seems to have done the trick thank you. Will check through properly tomorrow and confirm. Cheers.

**Reply by travisfranck** [reply]
Glad you're making progress!

**Reply by Karaka** [reply]
Sorry for delay getting back @chrispcampbell  @travisfranck ... waylaid with work! I guess thats not all bad.
So --genc all good with your change @chrispcampbell to the directData syntax in the spec.json file ...thank you.

Bouyed on by that success, I changed the launch.json command to generate html ... my goal being to build a webapp. 
In the launch command below, I simply swapped --genc for --genhtml...leaving all other config variables the same:
![Screenshot 2021-12-20 at 15 23 05](https://user-images.githubusercontent.com/7798808/146794330-da15c947-50e7-4acc-bc0d-9f125dcf966b.png)

But it seems like the directData section of the spec file isnt picked up for the --genhtml option? So I'm back to getting the "Direct data workbook not found" error (see screenshot of terminal output below).  Oddly, something is upsetting the parsing of the model variables too...with two errant refId's?
Maybe I'm not setting up the launch command for --genhtml correctly?

![Screenshot 2021-12-20 at 15 21 26](https://user-images.githubusercontent.com/7798808/146795678-93d05f4c-050f-4338-a0dd-95a3dfe5ceb3.png)

**Reply by Karaka** [reply]
@chrispcampbell @travisfranck
Ignore the errant refid's noted above gents.  I hadnt set up the graphs.csv config file correctly.  sorry. an amateur mistake. that is now working but I'm still getting the Direct data workbook not found error.
I notice in sde-generate.js that the directData section of my spec file is stripped out of opts.spec for the genhtml option (at lines 84-87) so spec.directData is false when the programme tries to read the xlsx file into directData.  not sure how material?

grateful for any thoughts...cheers.
